go/adbc/driver/snowflake/bulk_ingestion.go

```diff
@@ -0,0 +1,549 @@
+		if err != io.EOF {
+			return err
+		}
+		return nil
+	})
+
+	// Create a temporary stage, we can't start uploading until it has been created
+	_, err = st.cnxn.cn.ExecContext(ctx, createTemporaryStageStmt, nil)
+	if err != nil {
+		return
+	}
+
+	// Start uploading the file to Snowflake
+	fileName := "0.parquet" // Only writing 1 file, so use same name as first file written by ingestStream() for consistency
+	err = uploadStream(ctx, st.cnxn.cn, r, fileName)
+	if err != nil {
+		return
+	}
+
+	// Parquet writing is already done if the upload finished, so we're just checking for any errors
+	err = g.Wait()
+	if err != nil {
+		return
+	}
+
+	// Load the uploaded file into the target table
+	_, err = st.cnxn.cn.ExecContext(ctx, copyQuery, []driver.NamedValue{{Value: st.targetTable}})
+	if err != nil {
+		return
+	}
+
+	// Check final row count of target table to get definitive rows affected
+	nrows, err = countRowsInTable(ctx, st.cnxn.sqldb, st.targetTable)
+	return
+}
+
+// ingestStream performs bulk ingestion of a RecordReader and returns the
+// number of rows affected.
+//
+// The RecordReader must already be bound by calling stmt.BindStream(), and will
+// be released and reset upon completion.
+func (st *statement) ingestStream(ctx context.Context) (nrows int64, err error) {
+	defer func() {
+		st.streamBind.Release()
+		st.streamBind = nil
+	}()
+	defer func() {
+		// Always check the resulting row count, even in the case of an error. We may have ingested part of the data.
+		ctx := context.Background() // TODO(joellubi): switch to context.WithoutCancel(ctx) once we're on Go 1.21
+		n, countErr := countRowsInTable(ctx, st.cnxn.sqldb, st.targetTable)
+		nrows = n
+
+		// Ingestion, row-count check, or both could have failed
+		// Wrap any failures as ADBC errors
+
+		// TODO(joellubi): simplify / improve with errors.Join(err, countErr) once we're on Go 1.20
+		if err == nil {
+			err = errToAdbcErr(adbc.StatusInternal, countErr)
+			return
+		}
+
+		// Failure in the pipeline itself
+		if errors.Is(err, context.Canceled) {
+			err = errToAdbcErr(adbc.StatusCancelled, err)
+		} else {
+			err = errToAdbcErr(adbc.StatusInternal, err)
+		}
+	}()
+
+	parquetProps, arrowProps := newWriterProps(st.alloc, st.ingestOptions)
+	g, gCtx := errgroup.WithContext(ctx)
+
+	// Read records into channel
+	records := make(chan arrow.Record, st.ingestOptions.writerConcurrency)
+	g.Go(func() error {
+		return readRecords(gCtx, st.streamBind, records)
+	})
+
+	// Read records from channel and write Parquet files in parallel to buffer pool
+	schema := st.streamBind.Schema()
+	pool := newBufferPool(int(st.ingestOptions.targetFileSize))
+	buffers := make(chan *bytes.Buffer, st.ingestOptions.writerConcurrency)
+	g.Go(func() error {
+		return runParallelParquetWriters(gCtx, schema, int(st.ingestOptions.targetFileSize), int(st.ingestOptions.writerConcurrency), parquetProps, arrowProps, pool.GetBuffer, records, buffers)
```

Comment:
split this line for readability?

Reply:
Sure, I split this and a few other long ones.